org.eclipse.lsp4e/src/org/eclipse/lsp4e/internal/CharsInputStream.java

```diff
@@ -0,0 +1,224 @@
+	}
+
+	private enum EncoderState {
+		ENCODING,
+		FLUSHING,
+		DONE
+	}
+
+	/** 512 surrogate character pairs */
+	private static final int DEFAULT_BUFFER_SIZE = 512;
+	private static final int EOF = -1;
+
+	private final int bufferSize;
+	private final CharBuffer charBuffer;
+	private final ByteBuffer byteBuffer;
+	private final CharsetEncoder encoder;
+	private EncoderState encoderState = EncoderState.ENCODING;
+
+	private int charIndex = 0;
+	private final CharsSupplier chars;
+	private final IntSupplier charsLength;
+
+	public CharsInputStream(final CharSequence chars) {
+		this(chars, null);
+	}
+
+	public CharsInputStream(final CharSequence chars, final @Nullable Charset charset) {
+		this(chars, charset, DEFAULT_BUFFER_SIZE);
+	}
+
+	public CharsInputStream(final CharSequence chars, final @Nullable Charset charset, final int bufferSize) {
+		this(chars::charAt, chars::length, charset, bufferSize);
+	}
+
+	public CharsInputStream(final CharsSupplier chars, final IntSupplier charsLength) {
+		this(chars, charsLength, null);
+	}
+
+	/**
+	 * @param chars function to access indexed chars.
+	 * @param charsLength function to get the number of indexed chars provided by the <code>chars</code> parameter.
+	 */
+	public CharsInputStream(final CharsSupplier chars, final IntSupplier charsLength, final @Nullable Charset charset) {
+		this(chars, charsLength, charset, DEFAULT_BUFFER_SIZE);
+	}
+
+	/**
+	 * @param chars function to access indexed chars.
+	 * @param charsLength function to get the number of indexed chars provided by the <code>chars</code> parameter.
+	 * @param bufferSize number of surrogate character pairs to encode at once.
+	 */
+	public CharsInputStream(final CharsSupplier chars, final IntSupplier charsLength, final @Nullable Charset charset, final int bufferSize) {
+		if (bufferSize < 1)
+			throw new IllegalArgumentException("[bufferSize] must be 1 or larger"); //$NON-NLS-1$
+		encoder = (charset == null ? StandardCharsets.UTF_8 : charset).newEncoder();
+
+		this.bufferSize = bufferSize;
+		charBuffer = CharBuffer.allocate(bufferSize * 2); // buffer for 2 chars (high/low surrogate)
+		byteBuffer = ByteBuffer.allocate(bufferSize * 4); // buffer for one UTF character (up to 4 bytes)
+		byteBuffer.flip();
+		charBuffer.flip();
+
+		this.chars = chars;
+		this.charsLength = charsLength;
+	}
+
+	@Override
+	public int available() {
+		final int remaining = byteBuffer.remaining();
+		return remaining == 0 ? charsLength.getAsInt() - charIndex : remaining;
+	}
+
+	public Charset getCharset() {
+		return encoder.charset();
+	}
+
+	private boolean flushEncoder() throws IOException {
+		if (encoderState == EncoderState.DONE)
+			return false;
+
+		if (encoderState == EncoderState.ENCODING) {
+			encoderState = EncoderState.FLUSHING;
+		}
+
+		// flush
+		byteBuffer.clear();
+		final CoderResult result = encoder.flush(byteBuffer);
+		byteBuffer.flip();
+
+		if (result.isOverflow()) // byteBuffer too small
+			return true;
```

Comment:
why not `result.throwException()`?

Reply:
because we can handle it. it just means that the bytebuffer we use for char processing is to small in the flushing phase when the encoder may return (based on the charset) other additional bytes.